src/Microsoft.AspNetCore.Blazor.Browser.JS/src/Interop/InvokeJavaScriptFunctionWithJsonMarshalling.ts

```diff
@@ -0,0 +1,66 @@
+import { platform } from '../Environment';
+import { System_String } from '../Platform/Platform';
+import { getRegisteredFunction } from './RegisteredFunction';
+import { invokeDotNetMethod, MethodOptions, InvocationResult } from './InvokeDotNetMethodWithJsonMarshalling';
+import { getElementByCaptureId } from '../Rendering/ElementReferenceCapture';
+import { System } from 'typescript';
+import { error } from 'util';
+
+const elementRefKey = '_blazorElementRef'; // Keep in sync with ElementRef.cs
+
+export function invokeWithJsonMarshalling(identifier: System_String, ...argsJson: System_String[]) {
+  let result: InvocationResult;
+  const identifierJsString = platform.toJavaScriptString(identifier);
+  const args = argsJson.map(json => JSON.parse(platform.toJavaScriptString(json), jsonReviver));
+
+  try {
+    result = { succeeded: true, result: invokeWithJsonMarshallingCore(identifierJsString, ...args) };
+  } catch (e) {
+    result = { succeeded: false, message: e.message };
+  }
+
+  const resultJson = JSON.stringify(result);
+  return platform.toDotNetString(resultJson);
+}
+
+function invokeWithJsonMarshallingCore(identifier: string, ...args: any[]) {
+  const funcInstance = getRegisteredFunction(identifier);
+  const result = funcInstance.apply(null, args);
+  if (result !== null && result !== undefined) {
+    return result;
+  } else {
+    return null;
+  }
+}
+
+const invokeDotNetCallback: MethodOptions = {
+  type: {
+    assembly: 'Microsoft.AspNetCore.Blazor.Browser',
+    name: 'Microsoft.AspNetCore.Blazor.Browser.Interop.TaskCallback'
+  },
+  method: {
+    name: 'InvokeTaskCallback'
+  }
+};
+
+export function invokeWithJsonMarshallingAsync<T>(identifier: string, callbackId: string, ...argsJson: string[]) {
+  const result = invokeWithJsonMarshallingCore(identifier, ...argsJson) as Promise<any>;
```

Comment:
I don't think this accounts for synchronous exceptions.

The usual way to unify synchronous and asynchronous exceptions is something like:

```javascript
return new Promise(_ => {
    return someMethodThatReturnsAPromiseOrThrowsSynchronously();
});
```

... since the `Promise` constructor will catch any synchronous exceptions and then returns a promise in a pre-rejected state.

Reply:
Would it be possible to have test coverage for synchronous exceptions (both for .NET->JS calls and vice-versa)? Doesn't need to cover all the billions of possible combinations of argument types, just one example would be fine!

Reply:
This does account for sync exceptions as it's called as part of Invoke which captures the exception thrown synchronously.

I have explicit tests for this.

Reply:
Also, the .NET-side `Untrack` is never called in success cases either, which I assume is still to be implemented

Reply:
`InvokeAsync<T>` calls invokeWithJsonMarshalling through `Invoke<T>` already handles exceptions on the JavaScript side and returns an `InvocationResult<T>`

Reply:
As of your latest update, it now looks like `Untrack` is called in the success case (or async exception case), but still not in the synchronous exception case.